astropy/coordinates/astrom_provider.py

```diff
@@ -0,0 +1,220 @@
+    '''
+    The default provider for astrometry values.
+    A utility class to extract the necessary arguments for
+    erfa functions from frame attributes.
+    '''
+
+    @staticmethod
+    def apci(frame):
+        jd1_tt, jd2_tt = get_jd12(frame.obstime, 'tt')
+        cip = get_cip(jd1_tt, jd2_tt)
+        earth_pv, earth_heliocentric = prepare_earth_position_vel(frame.obstime)
+        return erfa.apci(jd1_tt, jd2_tt, earth_pv, earth_heliocentric, *cip)
+
+    @staticmethod
+    def apcs(frame):
+        jd1_tt, jd2_tt = get_jd12(frame.obstime, 'tt')
+        pv = pav2pv(
+            frame.obsgeoloc.get_xyz(xyz_axis=-1).value,
+            frame.obsgeovel.get_xyz(xyz_axis=-1).value
+        )
+        earth_pv, earth_heliocentric = prepare_earth_position_vel(frame.obstime)
+        return erfa.apcs(jd1_tt, jd2_tt, pv, earth_pv, earth_heliocentric)
+
+    @staticmethod
+    def apio13(frame):
+        lon, lat, height = frame.location.to_geodetic('WGS84')
+
+        jd1_utc, jd2_utc = get_jd12(frame.obstime, 'utc')
+        dut1utc = get_dut1utc(frame.obstime)
+
+        return erfa.apio13(
+            jd1_utc, jd2_utc, dut1utc,
+            lon.to_value(u.radian),
+            lat.to_value(u.radian),
+            height.to_value(u.m),
+            *get_polar_motion(frame.obstime),
+            # all below are already in correct units because they are QuantityFrameAttribues
+            frame.pressure.value,
+            frame.temperature.value,
+            frame.relative_humidity.value,
+            frame.obswl.value,
+        )
+
+
+class InterpolatingAstromProvider(AstromProvider):
+    '''
+    A provider for astrometry values that does not call erfa
+    for each individual timestamp but interpolates the linearly
+    between support points.
+
+    This can dramatically speed up coordinate transformations when
+    obstime is an array of many values (factors of 10 to > 100 depending
+    on the selected resolution and the time range of the values).
+    '''
+
+    @u.quantity_input(time_resolution=u.day)
+    def __init__(self, time_resolution):
+        if time_resolution.to_value(u.ms) < 1:
+            warnings.warn(
+                f'Using {self.__class__.__name__} with `time_resolution`'
+                ' below 1ms might not improve performance',
+                AstropyWarning
+            )
+        self.mjd_resolution = time_resolution.to_value(u.day)
+
+    def _get_support_points(self, obstime):
+        '''
+        Calculate support points for the interpolation.
+        This includes the minimum and maximum timestamp in obstime
+        to improve precision when the time resolution is wider than
+        the time range or only a single value is requested.
+        '''
+        mjd_scaled = np.ravel(obstime.mjd / self. mjd_resolution)
+
+        # ndmin=1 needed in case only a single coordinate is transformed
+        mjd_lower = np.array(np.floor(mjd_scaled), ndmin=1, copy=False)
```

Comment:
I could imagine, that a simple cast to int64 would be faster. Also, mjd_upper is redundant, isn't it? Could just append `mjd_lower[-1] + 1` to the array, if I'm not mistaken. (Would working with integers speed up the `np.unique` function? I guess not.)

Reply:
> Also, mjd_upper is redundant, isn't it? Could just append `mjd_lower[-1] + 1` to the array, if I'm not mistaken.
Strike that, it would not work for inhomogeneous data. However, `mjd_upper = mjd_lower + 1` should be ok.

Reply:
I'll check performance of these variants. But this anyhow does not really contribute to the performance costs here.

Reply:
```

In [1]: import numpy as np                                                                                                                    

In [2]: mjd = np.random.uniform(50000, 60000, 10000)                                                                                          

In [3]: %timeit np.ceil(mjd)                                                                                                                  
2.92 µs ± 47 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit np.int64(mjd)                                                                                                                 
5.47 µs ± 6.99 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit np.floor(mjd)                                                                                                                 
2.83 µs ± 12.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: lower = np.floor(mjd)                                                                                                                 

In [7]: %timeit upper = lower + 1                                                                                                             
3.42 µs ± 13.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit upper = np.ceil(mjd)                                                                                                          
3.01 µs ± 14.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Performance seems best for what it is now, calling floor and ceil

Reply:
Indeed, casting seems to be quite slow to ufunc (e.g., `x.astype(bool)` is slower than `x != 0`). Though as you wrote before, this part will not dominate execution time anyway, so clarity before anything else!